### DISCUSSION TITLE
Some "plugins" to add to the community

### DISCUSSION THREAD

**Discussion by Brian-McBride** [colinhacks/zod#535] [Show and tell]
## NPM Packages:

- ### [@anatine/zod-openapi](https://github.com/anatine/zod-plugins/tree/main/libs/zod-openapi)
  - Converts a Zod schema into an OpenAPI 3.0 `SchemaObject`
  - Leverages [openapi3-ts](https://www.npmjs.com/package/openapi3-ts) for typings and builders

- ### [@anatine/zod-mock](https://github.com/anatine/zod-plugins/tree/main/libs/zod-mock)
  - Generates a mock objet for testing.
  - Fake data generated from the peer dependency [faker.js](https://www.npmjs.com/package/faker) 

- ### [@anatine/zod-nestjs](https://github.com/anatine/zod-plugins/tree/main/libs/zod-nestjs)
  - Helper tooling to use Zod in [NestJS](https://nestjs.com/).
  - Patch for [@nestjs/swagger](https://docs.nestjs.com/openapi/introduction) to use `@anatine/zod-openapi` to display schemas.

---- 

## Goals

I like where `tRPC` is going. I hope to bring in the same ideas around a cleaner client/server connection. I'm concerned about conflicting types (server-based data vs client data) or unexpected imports into the client's compiled code. 

I'd like to see a pattern with a shared lib of the base Zod objects. Then libs that wrap those into server structures (express, fastify, etc...) and client (axios, fetch, etc...) without _any_ cross imports between the two.

My other use case is OpenAPI. There are just too many gateways or other supporting infrastructure that ask for an API Schema. These libs are following a code-first approach. In the case of NestJS - it is very easy to set up a script to generate the swagger doc, save it as a file or upload it to an API Gateway in the CI/CD process.

----

## Credits

These are not unique libs. I went to use some of the other libs in the Zod community section and found some were outdated (broken) and others just didn't quite cover my use cases. 

- `zod-dto` : Extensive use and inspiration for working with NestJS
- `express-zod-api` : Offered great insight into parsing Zod.

**Comment by Brian-McBride** [comment]
@colinhacks 
I just finished an update on these plugins. 
Still hoping to get them added to the community. 

The mock is pretty cool, in that it can be pretty contextual faker.js behind it.

**Reply by StefanTerdell** [reply]
@Brian-McBride Please create a PR with links added to the readme :)
